owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/shares/CreateRemoteShareOperation.kt

```diff
@@ -81,90 +91,120 @@ class CreateRemoteShareOperation(
+            val expirationDate = Calendar.getInstance()
+            expirationDate.timeInMillis = expirationDateInMillis
+            val formattedExpirationDate = dateFormat.format(expirationDate.time)
+            formBodyBuilder.add(PARAM_EXPIRATION_DATE, formattedExpirationDate)
+        }
+
+        if (publicUpload) {
+            formBodyBuilder.add(PARAM_PUBLIC_UPLOAD, publicUpload.toString())
+        }
+        if (password.isNotEmpty()) {
+            formBodyBuilder.add(PARAM_PASSWORD, password)
+        }
+        if (RemoteShare.DEFAULT_PERMISSION != permissions) {
+            formBodyBuilder.add(PARAM_PERMISSIONS, permissions.toString())
+        }
+
+        return formBodyBuilder
+    }
 
-            val requestUri = client.baseUri
-            val uriBuilder = requestUri.buildUpon()
-            uriBuilder.appendEncodedPath(ShareUtils.SHARING_API_PATH)
+    override fun run(client: OwnCloudClient): RemoteOperationResult<ShareResponse> {
+        val requestUri = buildRequestUri(client.baseUri)
 
-            val postMethod = PostMethod(URL(uriBuilder.build().toString()), formBodyBuilder.build())
+        val formBodyBuilder = createFormBodyBuilder()
 
-            postMethod.setRequestHeader(HttpConstants.CONTENT_TYPE_HEADER, HttpConstants.CONTENT_TYPE_URLENCODED_UTF8)
-            postMethod.addRequestHeader(OCS_API_HEADER, OCS_API_HEADER_VALUE)
+        val postMethod = PostMethod(URL(requestUri.toString()), formBodyBuilder.build()).apply {
+            setRequestHeader(HttpConstants.CONTENT_TYPE_HEADER, HttpConstants.CONTENT_TYPE_URLENCODED_UTF8)
+            addRequestHeader(OCS_API_HEADER, OCS_API_HEADER_VALUE)
+        }
 
+        return try {
             val status = client.executeHttpMethod(postMethod)
+            val response = postMethod.getResponseBodyAsString()
 
-            val parser = ShareToRemoteOperationResultParser(
-                ShareXMLParser()
-            )
-
-            if (isSuccess(status)) {
-                parser.oneOrMoreSharesRequired = true
-                parser.ownCloudVersion = client.ownCloudVersion
-                parser.serverBaseUri = client.baseUri
-                result = parser.parse(postMethod.getResponseBodyAsString())
-
-                if (result.isSuccess && retrieveShareDetails) {
-                    // retrieve more info - POST only returns the index of the new share
-                    val emptyShare = result.data.shares[0]
-                    val getInfo = GetRemoteShareOperation(
-                        emptyShare.id
-                    )
-                    result = getInfo.execute(client)
-                }
-
+            if (!isSuccess(status)) {
+                onResultUnsuccessful(postMethod, response, status)
             } else {
-                result = parser.parse(postMethod.getResponseBodyAsString())
+                onRequestSuccessful(response)
             }
```

Comment:
Normally, I like to have this order instead of having an `if(!whatever) { } else { }`

           if (isSuccess(status)) {
                onRequestSuccessful(response)
            } else {
                onResultUnsuccessful(postMethod, response, status)
            }

Reply:
Fixed :)